### PR TITLE
feat: Add LocationHelper for managing user location and permissions

### PR DESCRIPTION
**Pull request by haidyhesham95**
- Implemented saving and retrieving location using SharedPreferences.
- Added methods to check location services and request permissions.
- Enhanced error handling and fallback for location-related operations.
- Included geolocation and reverse geocoding functionality.           
                                                                                                      
WhatsApp Image 2024-12-18 at 23 31 38_712c50a9
WhatsApp Image 2024-12-18 at 23 31 38_adea7265
WhatsApp Image 2024-12-18 at 23 31 38_4e566cfd
WhatsApp Image 2024-12-18 at 23 31 38_f34dfe1f

### CODE REVIEW

File: lib/core/services/shared_preference/location_helper.dart [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import 'package:flutter_gen/gen_l10n/app_localizations.dart';
+
+import '../../../features/auth/data/data_sources/contracts/offline_data_source.dart';
+class LocationHelper {
+  final OfflineDataSource offlineDataSource = getIt<OfflineDataSource>();
+
+  Future<void> saveLocation(String location) async {
+    try {
+      SharedPreferences prefs = await SharedPreferences.getInstance();
+      String? token = await offlineDataSource.getToken();
```
Comment (Marco-Nagy) [inline lib/core/services/shared_preference/location_helper.dart]: you don't need to concatenate location kye with token

i prefer just save location value with location key

File: lib/features/home_screen/presentation/widgets/location _widget.dart [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+      return;
+    }
+
+    try {
+      Position position = await _locationHelper.getCurrentLocation(context);
+      String address = await _locationHelper.getAddressFromLatLng(
+          position.latitude, position.longitude, context);
+
+      await _locationHelper.saveLocation(address);
+
```
Comment (Marco-Nagy) [inline lib/features/home_screen/presentation/widgets/location _widget.dart]: please refactor set state functions at cart view model

**Review by Marco-Nagy — CHANGES_REQUESTED** [review]
عاااش يا هيدى 
شغل عالى عالى الصراحة
بس محتاجين نهندل ال functions فى ال view model افضل ك clean code
و اعتقد محتاجين نضيف comments as documentation at location helper class and location widget

File: lib/features/auth/presentation/login/viewModel/login_view_model_cubit.dart [hunk 3]
```diff
@@ -32,6 +35,21 @@ class LoginViewModel extends Cubit<LoginViewModelState> {
 
   Future<void> _login(LoginAction action, BuildContext context) async {
     emit(LoginViewModelLoading());
+    // Check location permission
+    bool isServiceEnabled = await _locationHelper.isLocationServiceEnabled();
+    if (!isServiceEnabled) {
+      emit(LoginViewModelError(ErrorHandler.handle(
+          Exception(AppLocalizations.of(context)!.location_services_are_disabled))));
+      return;
+    }
```
Comment (Marco-Nagy) [inline lib/features/auth/presentation/login/viewModel/login_view_model_cubit.dart:42]: التعديل ده محتاجين نعمله فى function منفصله و نعملها call فى ال login bloc listener at loading case

File: lib/features/cart/presentation/widgets/cart_view_body.dart [hunk 4]
```diff
@@ -14,15 +14,43 @@ import 'package:flutter/material.dart';
...
+
+  @override
+  void initState() {
+    super.initState();
+    _loadLocation();
+  }
+
+  Future<void> _loadLocation() async {
+    String? savedLocation = await _locationHelper.getSavedLocation();
+    setState(() {
```
Comment (Marco-Nagy) [inline lib/features/cart/presentation/widgets/cart_view_body.dart]: محتاجين ننقل الfunction دى لل view model و نعملها call هنا بس

File: lib/features/home_screen/presentation/widgets/location _widget.dart [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    return BlocProvider(
+      create: (context) {
+        final cubit = LocationCubit(LocationHelper());
+        cubit.loadSavedLocation(context);
+        return cubit;
+      },
+      child: BlocBuilder<LocationCubit, LocationState>(
+        builder: (context, state) {
+          if (state is LocationLoading) {
+            return Center(
```
Comment (Marco-Nagy) [inline lib/features/home_screen/presentation/widgets/location _widget.dart]: الاحسن return awesome snackbar loading

**Review by Marco-Nagy — CHANGES_REQUESTED** [review]
محتاجين نعدل حاجات بسيطة بس الله ينور شغل عاالى جدا 👏  👏  👏  👏  👏  👏

File: lib/features/home_screen/presentation/widgets/location _widget.dart (hunk 5 above)
Comment (haidyhesham95) [reply lib/features/home_screen/presentation/widgets/location _widget.dart]: مش هتبقي حلوه كده احسن

File: lib/features/auth/presentation/login/viewModel/login_view_model_cubit.dart (hunk 3 above)
Comment (haidyhesham95) [reply lib/features/auth/presentation/login/viewModel/login_view_model_cubit.dart:42]: عدلته

File: lib/features/cart/presentation/widgets/cart_view_body.dart (hunk 4 above)
Comment (haidyhesham95) [reply lib/features/cart/presentation/widgets/cart_view_body.dart]: عدلته

**Review by Marwan2595 — APPROVED** [review]

**Review by Marco-Nagy — APPROVED** [review]

**Pull request merged**